src/queue/descriptor_chain.rs

```diff
@@ -178,3 +184,815 @@ impl<M: GuestAddressSpace> Iterator for DescriptorChainRwIter<M> {
+    }
+
+    fn split_at(&mut self, offset: usize) -> Result<DescriptorChainConsumer<'a>> {
+        let (at, offset) = self
+            .position(offset)
+            .ok_or(Error::DescriptorChainSplitOOB(offset))?;
+
+        let (tail, head) = self.buffers[at].split_at(offset).unwrap();
+        // let's make a copy for self
+        let mut buffers = self.buffers[self.index..at].to_vec();
+        buffers.push(tail);
+        std::mem::swap(&mut self.buffers, &mut buffers);
+        self.index = 0;
+        // update the returned head
+        buffers[at] = head;
+        Ok(DescriptorChainConsumer {
+            buffers,
+            bytes_consumed: 0,
+            index: at,
+        })
+    }
+}
+
+fn collect_desc_chain_buffers<M: GuestAddressSpace>(
+    mem: &GuestMemoryMmap,
+    desc_chain_iter: DescriptorChainRwIter<M>,
+) -> Result<Vec<VolatileSlice<'_>>> {
+    let mut total_len: usize = 0;
+    let mut buffers = Vec::new();
+    for desc in desc_chain_iter {
+        // Verify that summing the descriptor sizes does not overflow.
+        // This can happen if a driver tricks a device into reading more data than
+        // fits in a `usize`.
+        total_len = total_len
+            .checked_add(desc.len() as usize)
+            .ok_or(Error::DescriptorChainOverflow)?;
+
+        let mut len = desc.len() as usize;
+        let mut addr = desc.addr();
+        while len > 0 {
+            let region = mem.find_region(addr).ok_or(Error::FindMemoryRegion)?;
+            let offset = addr
+                .checked_sub(region.start_addr().raw_value())
+                .unwrap()
+                .raw_value() as usize;
+            let buf_len = cmp::min(region.deref().len() - offset, len);
+            let buf = region
+                .deref()
+                .get_slice(offset, buf_len)
+                .map_err(Error::VolatileMemoryError)?;
+            buffers.push(buf);
+            len -= buf_len;
+            addr = addr
+                .checked_add(buf_len as _)
+                .ok_or(Error::DescriptorChainOverflow)?;
+        }
+    }
+
+    Ok(buffers)
+}
+
+/// Provides high-level interface over the sequence of memory regions
+/// defined by readable descriptors in the descriptor chain.
+///
+/// Note that virtio spec requires driver to place any device-writable
+/// descriptors after any device-readable descriptors (2.6.4.2 in Virtio Spec v1.1).
+/// Reader will skip iterating over descriptor chain when first writable
+/// descriptor is encountered.
+#[derive(Clone, Debug)]
+pub struct Reader<'a> {
+    buffer: DescriptorChainConsumer<'a>,
+}
+
+impl<'a> Reader<'a> {
+    /// Construct a new Reader wrapper over `desc_chain`.
+    pub fn new<M: GuestAddressSpace>(
+        mem: &'a GuestMemoryMmap,
+        desc_chain: DescriptorChain<M>,
+    ) -> Result<Reader<'a>> {
+        Ok(Reader {
+            buffer: DescriptorChainConsumer {
+                buffers: collect_desc_chain_buffers(mem, desc_chain.readable())?,
```

Comment:
I've tried experimenting a bit, and I encountered the following question: what's the usage pattern for  the `Reader` and `Writer` when a descriptor chain has both device readable and writable descriptors? For example, if we start by creating a `Reader`, this uses `desc_chain.readable()`, which filters out all non-readable descriptors. What is the standard approach to reading bytes from the readable segments of a descriptor chain, and then writing stuff to the writable descriptors of the same chain using these abstractions?

Reply:
Honestly, I don't think this is very common. In all cases, Reader/Writer is not meant to be the only high-level abstraction, if you have some weird case like this, then perhaps you shouldn't use it.

Reply:
To give an example, virtio block device operation (described [here](https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-2500006)) requires something like this. I think it can be solved with some tinkering on top of the current proposal; I'll try a few things out and return with a proposal in a couple of days tops. Meanwhile, if you can share some more details around the "why the need for two separate memory handles" above, that would be of great help as well.

Reply:
yes, I am working on it. I am not very familiar with the vmm crates..

Reply:
I'll be happy to provide more context around the current state of things (or simply bounce ideas) if it helps. You can find myself and other folks from around here on the rust-vmm Slack workspace.